### PR TITLE
GCG-34 Write class to generate code for all supported jsons

### PR DESCRIPTION
**Pull request by prncoprs**
add the class which can generate code from query numbers in the test.

According to the query number, generate code for all supported jsons.

### CODE REVIEW

**Comment by prncoprs** [comment]
@DanishAlsayed Could you plz help me review it?

**Review by DanishAlsayed — COMMENTED** [review]
Did you try and run this code? Why is the `generated-code` folder still there?

**Review by DanishAlsayed — COMMENTED** [review]
We should also provide the option to run the main function directly and not have to run codegen jar. And also move the logic to compile `codegen` from `IntegrationTest` to here and compile codegen everytime the tester wants to test with codegen jar, just like `IntegrationTest`

**Comment by DanishAlsayed** [comment]
Please eliminate all compiler warnings

File: codegen/src/test/java/GenerateCodeTest.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+        initQueryArrayList();
+
+        // check the child folder of the resource folder
+        File[] resourceFolderFiles = resourceFolderFile.listFiles();
+        for (File file : resourceFolderFiles) {
+            if (!file.isDirectory()) {
+                continue;
+            }
+            if (file.getName().charAt(0) != 'q') {
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:36]: fixed in the next commit.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            }
+            int queryNum;
+            try {
+                queryNum = Integer.parseInt(file.getName().substring(1));
+            } catch (NumberFormatException e) {
+                System.out.println(file.getName() + " is not a query name.");
+                continue;
+            }
+            // TPC-H has only 1~22 queries.
+            if (queryNum < 1 || queryNum > 22) {
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:47]: fixed in the next commit.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                System.out.println("TPC-H does not contain query " + String.valueOf(queryNum) + ".");
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:48]: removed this check

**Comment by prncoprs** [comment]
> Did you try and run this code? Why is the `generated-code` folder still there?

Removed the `generated-code` after copy the scala file to query folders.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            if (validateQueryJsonFileExist(queryNum)) {
+                queryArrayList.add(queryNum);
+            }
+        }
+    }
+
+    /*
+     * init the queryArrayList
+     */
+    private static void initQueryArrayList() {
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:60]: removed this function.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        while ((output = stdInput.readLine()) != null) {
+            System.out.println(output);
+        }
+        while ((output = stdError.readLine()) != null) {
+            System.out.println("ERROR:" + output);
+        }
+    }
+
+    private void copyGeneratedCodeToQueryFolder(int queryIdx) throws IOException {
+        File queryFolderFile = new File(RESOURCE_FOLDER + File.separator + "q" + String.valueOf(queryIdx));
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:128]: extract a new method as `private File getQueryFolderFile(int queryIdx)` .

File: codegen/src/test/java/GenerateCodeTest.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    private void runCommand(String command_str) throws IOException, InterruptedException {
+        Process process = Runtime.getRuntime().exec(command_str);
+        process.waitFor();
+
+        System.out.println("run command: " + command_str);
+
+        BufferedReader stdInput = new BufferedReader(new InputStreamReader(process.getInputStream()));
+        BufferedReader stdError = new BufferedReader(new InputStreamReader(process.getErrorStream()));
+
+        String output = null;
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:118]: removed the `=null`.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            System.out.println(output);
+        }
+        while ((output = stdError.readLine()) != null) {
+            System.out.println("ERROR:" + output);
+        }
+    }
+
+    private void copyGeneratedCodeToQueryFolder(int queryIdx) throws IOException {
+        File queryFolderFile = new File(RESOURCE_FOLDER + File.separator + "q" + String.valueOf(queryIdx));
+        File generatedCodeSourceFolder = new File(queryFolderFile.getAbsolutePath() + File.separator
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:129]: extract a new method as `private File getQueryFolderFile(int queryIdx)` .

File: codegen/src/test/java/GenerateCodeTest.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            File dstFile = new File(queryFolderFile.getAbsolutePath() + File.separator + file.getName());
+            if (dstFile.exists()) {
+                dstFile.delete();
+            }
+            Files.copy(file.toPath(), dstFile.toPath());
+        }
+    }
+
+    private void removeGeneratedCodeFolder(int queryIdx) throws Exception {
+        File queryFolderFile = new File(RESOURCE_FOLDER + File.separator + "q" + String.valueOf(queryIdx));
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:143]: changed the method `deleteFolder()` to `FileUtils.deleteDirectory()`.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    private static ArrayList<Integer> queryArrayList = null;
+
+    private static final String RESOURCE_FOLDER = new File("src" + File.separator + "test" + File.separator + "resources").getAbsolutePath();
+    private final String CODEGEN_JAR_PATH = new File(CodeGen.class.getProtectionDomain().getCodeSource().getLocation().getPath()).getParentFile().getAbsolutePath() + File.separator + "codegen-1.0-SNAPSHOT.jar";
+
+    @BeforeClass
+    public static void getQueryArrayList() {
+        // check if resource folder exists and is a directory
+        File resourceFolderFile = new File(RESOURCE_FOLDER);
+        assertTrue(resourceFolderFile.exists());
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:25]: changed this to if conditions and early returns with exceptions.

File: codegen/src/test/java/GenerateCodeTest.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        for (int queryIdx : queryArrayList) {
+            removeGeneratedCodeFolder(queryIdx);
+            generateCode(queryIdx);
+            copyGeneratedCodeToQueryFolder(queryIdx);
+        }
+    }
+
+    private void generateCode(int queryIdx) throws IOException, InterruptedException {
+        File queryFolderFile = new File(RESOURCE_FOLDER + File.separator + "q" + String.valueOf(queryIdx));
+        File queryJsonFile = new File(queryFolderFile.getAbsolutePath() + File.separator + "Q" + String.valueOf(queryIdx) + ".json");
```
[the remark this answers is not in the document]
Comment (prncoprs) [reply codegen/src/test/java/GenerateCodeTest.java:74]: changed to a new method to get `queryFolderFile`.

**Comment by prncoprs** [comment]
> We should also provide an option to run a given query, and not just all of them, in case we want to generate code code for a given query, not all of them.

In this class, there is a method called `private void generateCode(int queryIdx)`, which can generate code for a give query.

**Comment by prncoprs** [comment]
> We should also provide the option to run the main function directly and not have to run codegen jar. And also move the logic to compile `codegen` from `IntegrationTest` to here and compile codegen everytime the tester wants to test with codegen jar, just like `IntegrationTest`.

1. added the method `private void generateCodeUsingMainFunction(int queryIdx) throws Exception` to run the main function directly.
2. re-package the codegen.jar everytime to generate code.

**Comment by prncoprs** [comment]
@DanishAlsayed Could u plz help me review the latest two commits? Thx!

**Pull request merged**